CARDS-247: As a user, I can perform advanced searches across all data

veronikaslc opened the pull request:
Added [CARDS-1260](https://phenotips.atlassian.net/jira/software/c/projects/CARDS/issues/CARDS-1260): Add a new filter operator: CONTAINS

Commit:
CARDS-247: As a user, I can perform advanced searches across all data

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -133,8 +137,8 @@ public void doGet(final SlingHttpServletRequest request, final SlingHttpServletR
 
-    @SuppressWarnings({"checkstyle:ExecutableStatementCount"})
-    private String createQuery(final SlingHttpServletRequest request)
+    @SuppressWarnings({"checkstyle:ExecutableStatementCount", "checkstyle:NPathComplexity"})
```
Definitely not a good sign.

Commit:
CARDS-1260: Add a new filter operator: CONTAINS

Commit:
CARDS-1260: Add a new filter operator: CONTAINS
Addressed codereview comments

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,112 +462,220 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
+    }
 
-        // Build the filter conditionals by imposing conditions on the inner joined cards:Answer children
+    private String addSpeacialFilter(final String filter, final String value, final String comparator,
```
Typo: `addSpecialFilter`

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -170,75 +198,153 @@ private String createQuery(final SlingHttpServletRequest request)
+                filtersToPrefix
+            )
+        );
+        // For subject query case we parse existence together with other filters in previous step
```
False comment, the existence is not properly handled in the previous step, the joins are added, but the WHERE conditions are not.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -145,18 +150,41 @@ private String createQuery(final SlingHttpServletRequest request)
+                    "child", session);
+                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, null, filterempty,
+                    "empty", session);
+                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, null, filternotempty,
```
Why not this:

```suggestion
                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, filtersToPrefix, filterempty,
                    "empty", session);
                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, filtersToPrefix, filternotempty,
```

Without `filtersToPrefix`, there's a `NullPointerException` thrown at line 289.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,112 +462,220 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
+                    this.sanitizeField(value)
+                )
+            );
+        } else {
```
Missing  support for `is [not] empty`.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,112 +462,220 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
-                }
+                break;
+            default:
+                filterdata.append("");
```
This is useless, you don't need to add nothing.

```suggestion
```

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,112 +462,220 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
             }
         }
+        // Condition 2: the value must match
+        if ("contain".equals(comparator)) {
```
```suggestion
        if ("contains".equals(comparator)) {
```

sdumitriu reviewed (changes requested):
- bug: `is [not] empty` filters no longer work when filtering forms, and cause an exception when filtering subjects; to reproduce: create a Radiotherapy form with some text in the Comments field, add filters for `Radiotherapy/Comments is or is not empty`; cause: `getQuestionnairesMaps` is called with a `null` instead of `filtersToPrefix` two times
- fixing it for Subjects leads to a different problem: adding just an `isnotempty` filter doesn't actually put the `is not empty` condition it in the query, which leads to an answer join without any conditions on it, which leads to a resultset of very many identical subjects
- even after fixing this, it raises another possible problem: if the filters match the same subject multiple times, we'll get the same subject more than once in the result, e.g. if we have more Chemotherapy forms matching the filters, or if a repeatable section is repeated many times with each answer matching the filters; this is hard to fix in the query itself, since the support for `distinct` is broken in Oak; something to address another time
- bug: contains doesn't seem to work; create a Radiotherapy form, set `Comments = abc def ghi`, filter for `Radiotherapy/Comments contains def`; cause: a typo for `contains/contain`, 
- typos: it's `joins` not `joints`

veronikaslc committed:
CARDS-1260: Add a new filter operator: CONTAINS

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
CARDS-247: As a user, I can perform advanced searches across all data

Removed useless change

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

veronikaslc committed:
CARDS-247: As a user, I can perform advanced searches across all data

Avoiding NullPointerEsception

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

Commit:
CARDS-247: As a user, I can perform advanced searches across all data
Rename function - fixed typo

Commit:
CARDS-247: As a user, I can perform advanced searches across all data
Fixed typos

Commit:
CARDS-247: As a user, I can perform advanced searches across all data
Fixed is [not] null filter

Commit:
CARDS-247: As a user, I can perform advanced searches across all data
Avoided NPathComplexity

sdumitriu committed:
[cleanup] Fixed typo: joints -> joins

sdumitriu committed:
CARDS-247: As a user, I can perform advanced searches across all data
Bugfix: IS [NOT] EMPTY does not work when filtering subjects

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,120 +487,229 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
+    private String parseFilter(final String[] fields, final String[] values, final String[] types,
+        final String[] comparator, final String nodeType, Map<String, String> questionnairesToPrefix,
+        final Map<String, List<String>> questionnairesToFilters,
+        final Map<String, String> filtersToPrefix) throws IllegalArgumentException
```
The above Javadoc comment should be updated to describe what the new parameters do and additionally, a `@return` line should specify what this function returns.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,120 +487,229 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
+     * @param empties user input field names to assert the nonexistence of content for
+     * @param notempties user input field names to assert the existence of content for
      * @return JCR_SQL conditionals for the input
      */
```
Create a `@param` Javadoc line to describe what the new `filtersToPrefix` parameter does.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -475,10 +718,11 @@ private String parseExistence(final String[] empties, final String[] notempties)
-     * @param comparison unary comparitor to assert
+     * @param comparison unary comparator to assert
      * @return JCR_SQL conditionals for the input
      */
```
Create a `@param` Javadoc line to describe what the new `filtersToPrefix` parameter does.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,120 +487,229 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
 
-        // Also parse out multiple comparators
-        String[] comparators;
+        if (comparator != null && comparator.length != values.length) {
```
Should `fields.length` also be checked for equality with `comparator.length` and `values.length`, otherwise there is potential for an `ArrayIndexOutOfBoundsException` in the `Add filters on subject uuid, questionnaire or date separately before other filters` loop?

veronikaslc replied:
We do it in previous step `if (fields.length != values.length)`, so at this time both are equal size so this check is enough.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,120 +487,229 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
             // Use = as the default
             comparators = new String[fields.length];
+            Arrays.fill(comparators, "=");
+        }
```
The `comparators` vs. `comparator` naming convention can be confusing - especially given that both are lists. Why not rename `comparator` to `optionalComparators` and update the Javadoc to specify that these comparators are _optional_ in the sense that if `null` is passed the `=` comparator will be assumed?

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,120 +487,229 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
 
-        // Build the filter conditionals by imposing conditions on the inner joined cards:Answer children
+    private String addSpecialFilter(final String filter, final String value, final String comparator,
+        final String prefix, final String subjectSelector)
```
Some Javadoc commenting would be helpful to explain what the individual parameters do as well as what the _SpecialFilter_ is and how it differs from a _SingleFilter_.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -353,120 +487,229 @@ private String generateDateCompareQuery(String jcrVariable, String thisDayStr, S
+
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity"})
+    private String addSingleFilter(final String filter, final String value, final String comparator, final String type,
+            final String prefix)
```
Some Javadoc commenting would be helpful to explain what the parameters do as well as what the method does - for example, how the _SingleFilter_ is different from a _SpecialFilter_.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -170,75 +193,183 @@ private String createQuery(final SlingHttpServletRequest request)
+    private String processJoins(final String joinNodetype, Map<String, String> questionnairesToPrefix,
+        Map<String, List<String>> questionnairesToFilters, Map<String, String> filtersToPrefix,
+        final String[] filternames, final String[] filterempty, final String[] filternotempty,
+        Session session, final String nodeType)
```
Some Javadoc commenting would be helpful to explain what each of the parameters do as well as what is the _processing_ that is done to the _joins_ in this method.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -170,75 +193,183 @@ private String createQuery(final SlingHttpServletRequest request)
+
+    private void getQuestionnairesMaps(Map<String, String> questionnairesToPrefix,
+            Map<String, List<String>> questionnairesToFilters, Map<String, String> filtersToPrefix,
+            final String[] uuids, final String prefix, Session session)
```
Some Javadoc commenting would be helpful to explain what this method and its parameters do.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -170,75 +193,183 @@ private String createQuery(final SlingHttpServletRequest request)
+    }
+
+    private String createSubjectJoins(final String nodetype, Map<String, String> questionnairesToPrefix,
+            Map<String, List<String>> questionnairesToFilters, Map<String, String> filtersToPrefix)
```
Some Javadoc commenting would be helpful to explain what this method and its parameters do.

sashaandjic reviewed (approved):
tested

Commit:
CARDS-247: As a user, I can perform advanced searches across all data
Added comments

Commit:
CARDS-853: Transfer filters from Dashboard to forms page when clicking on Expand

Commit:
CARDS-853: Transfer filters from Dashboard to forms page when clicking on Expand
Fixed filter transfering

sdumitriu committed:
[misc] Speed up the FormRelatedSubjects editor by not descending down into form children

sdumitriu committed:
CARDS-247: As a user, I can perform advanced searches across all data
Major refactoring to simplify code and make it more extensible.

sdumitriu committed:
[misc] Fixed javadoc error

sashaandjic commented:
`Contains `operator works really so well, even in the pedigree!
Why didn't we add this operator to subjects?

sashaandjic commented:
> * typos: it's `joins` not `joints`

The best typo ever!

sashaandjic commented:
Subjects are listed multiple times if they have more than one form with the same condition. See the sample below:
![image](https://user-images.githubusercontent.com/23243244/131926426-2fa6a1e5-1669-4737-ab36-7fadbc778957.png)

sashaandjic commented:
Problem when doing filters `isEmpty` and `isNotEmpty` on `Subjects` and `Questionnaires`.
![image](https://user-images.githubusercontent.com/23243244/131926598-f86f91d7-5397-45b0-bc31-f43eee05669d.png)

sdumitriu commented:
I don't think is empty and is not empty are valid filters for subject and questionnaire. We should remove them.

sdumitriu commented:
Debugging why empty conditions on subject/questionnaire don't work, I think there really is no solution other than not allowing these filters. Rebasing to take into account #769 .

sdumitriu commented:
Duplicated subjects is a known problem that will be addressed later. Other than that, anything else wrong with this PR @sashaandjic ?

sashaandjic commented:
This filter :
![image](https://user-images.githubusercontent.com/23243244/132142840-a4ff9e3f-c8b1-42a1-b151-941f6b5b296f.png)

still translates to isEmpty
![image](https://user-images.githubusercontent.com/23243244/132142844-95e60ddf-c923-4387-8395-1768900328d0.png)

sashaandjic commented:
Everything else works as expected.

marta- commented:
> Everything else works as expected.
> Can we prevent `isEmpty `and `not isEmpty` in the selection dialog and display a warning message that the condition is invalid?

Or we could have a special case in the backend for these mandatory data (subject, questionnaire, creation date) and if the filter is empty return a json with zero results instead of an error.

sashaandjic commented:
> Or we could have a special case in the backend for these mandatory data (subject, questionnaire, creation date) and if the filter is empty return a json with zero results instead of an error.

Yes, this sounds as the proper way to handle the issue.

sdumitriu committed:
CARDS-853: Transfer filters from Dashboard to forms page when clicking on Expand
- allow reading filters and active tab from the URL on the dashboard too
- allow settings for both forms and subjects in the URL
- no more special case for the active tab, this setting also has a name in the URL

Commit:
CARDS-247: As a user, I can perform advanced searches across all data
Added a special case for isEmpty isNotEmpty filters

sashaandjic commented:
Modify filters shows nothing for `is not empty` filter
![image](https://user-images.githubusercontent.com/23243244/132564832-42408076-c898-44ba-bdcb-56f9ae1deaa7.png)

Also, `is not empty` filter for `Subjects` and `Questionnaires `always returns no results
![image](https://user-images.githubusercontent.com/23243244/132565002-842f8296-b1ea-4888-812a-4fba6a527e94.png)

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -82,47 +87,179 @@
-            Workspace workspace = session.getWorkspace();
-            QueryManager queryManager = workspace.getQueryManager();
+            // Check for special cases in request and return zero results if any
+            checkForSpecialEmptyFilter(request, filters, response);
```
If `checkForSpecialEmptyFilter` detects a case where zero results should be returned then it doesn't make sense to run any of the code below this check statement. Perhaps `void checkForSpecialEmptyFilter` should be changed to `boolean checkForSpecialEmptyFilter` so that `doGet` can return early if a special case arises where zero results need to be returned.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java:
```diff
@@ -82,47 +87,179 @@
+                    || CREATED_DATE_IDENTIFIER.equals(filter.name)) {
+                // Write the empty response
+                final Iterator<Resource> results = Collections.emptyIterator();
+                writeResponse(request, response, 0, 10, results);
```
For good practice, `limit` and `offset` should be taken from the `request` object instead of being hard-coded to `0` and `10` respectively. In this case, it's not too critical as we are always returning empty JSON so the _limit_ and _offset_ don't really matter but it would be nicer if the `limit` and `offset` values in the returned JSON actually matched the `limit` and `offset` URL parameters.

sdumitriu committed:
CARDS-247: As a user, I can perform advanced searches across all data
- Refactoring to return early
- Also use the fix when filtering forms
- IsNotEmpty should not return zero results, since anything would match

sdumitriu committed:
CARDS-247: As a user, I can perform advanced searches across all data
Bugfix: prevent IsEmpty subject/questionnaire filters from being created

sashaandjic reviewed (approved):
tested
